src/geeknn/ordination/_base.py

```diff
@@ -0,0 +1,282 @@
+
+        with Parallel(n_jobs=num_threads, backend="threading") as p:
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
+
+
+class Raw:
+    def __init__(self, k=1, max_duplicates=None):
+        self.k = k
+        self.max_duplicates = max_duplicates if max_duplicates is not None else 5
+
+    def transform_image(self, X_image: ee.Image) -> ee.Image:
+        """No-op placeholder."""
+        return X_image
+
+    def transform_fc(self, fc: ee.FeatureCollection) -> ee.FeatureCollection:
+        """No-op placeholder."""
+        return fc
+
+    @property
+    def k_nearest(self):
+        """Return the number of requested nearest neighbors plus max_duplicates
+        in order to account for non-independent samples.
+        """
+        return self.k + self.max_duplicates
+
+    def get_ids(self, fc: FeatureCollection, id_field: str) -> list[Any]:
+        """Return the IDs from the feature collection."""
+        return list(map(int, fc.aggregate_array(id_field)))
+
+    def _set_classifier(
+        self,
+        *,
+        fc: ee.FeatureCollection,
+        class_property: ee.String,
+        input_properties: ee.List,
+    ) -> None:
+        """Set the EE minimumDistance classifier for this estimator."""
+        self.clf = ee.Classifier.minimumDistance(
+            metric="euclidean",
+            kNearest=self.k_nearest,
+        ).train(
+            features=fc,
+            classProperty=class_property,
+            inputProperties=input_properties,
+        )
+
+    def train(
+        self, *, fc: ee.FeatureCollection, id_field: str, X_columns: list[str], **_
+    ):
+        """Train the estimator, storing the needed server-side objects used
+        in prediction.
+        """
+        self.id_field = ee.String(id_field)
+        self.X_columns = ee.List(X_columns)
+        self._set_classifier(
+            fc=fc, class_property=self.id_field, input_properties=self.X_columns
+        )
+        return self
+
+    def predict(self, X_image: ee.Image, mode: str = "CLASSIFICATION"):
+        """Predict the nearest neighbors for the given covariate (X) image."""
+        self.clf = self.clf.setOutputMode(ee.String(mode))
+
+        # Ensure the X_image band names match the X_columns
+        X_image = X_image.select(self.X_columns)
+
+        # Transform image
+        X_image = self.transform_image(X_image)
+
+        def get_neighbor_band_name(i):
+            return ee.String("NN").cat(ee.Number(i).int().format())
+
+        band_names = ee.List.sequence(1, self.k).map(get_neighbor_band_name)
+
+        return (
+            X_image.classify(classifier=self.clf)
+            .toArray()
+            .arraySlice(0, 0, self.k)
+            .arrayFlatten([band_names])
+        )
+
+    def predict_fc(self, fc: ee.FeatureCollection, colocation_obj=None):
```

Comment:
I've gone back and forth on whether or not this is a terrible suggestion, but you could consider using a [single dispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) for the `predict` and `transform` methods, so they would accept either an `Image` or `FeatureCollection` and internally dispatch to the correct method. It makes the public API a little simpler, but possibly also more confusing...

Reply:
This is a really cool idea - I've never dived deeply enough into `functools` to know it was there (takes me back to my dinosaur C days where varying the arguments creates different functions).  It looks like it's a pretty simple fix if we have an initial argument with the rest being keyword arguments:
```
from functools import singledispatchmethod

class Raw:
    @singledispatchmethod
    def predict(self, arg, **kwargs):
        raise NotImplementedError

    @predict.register
    def _(self, X_image: ee.Image, mode: str = "CLASSIFICATION"):
        ...

    @predict.register
    def _(self, fc: ee.FeatureCollection, colocation_obj=None):
        ...

fc_prediction = model.predict(fc, colocation_obj=colocation_obj)
img_prediction = model.predict(X_image, mode="CLASSIFICATION")
```
But then it doesn't look like you can call them like this anymore:
```
fc_prediction = model.predict(fc=fc, colocation_obj=colocation_obj)
img_prediction = model.predict(X_image=X_image, mode="CLASSIFICATION")
```
From your perspective, is the first way a more natural way to call this?  What were your reservations about suggesting this initially?

Reply:
> takes me back to my dinosaur C days where varying the arguments creates different functions

Exactly, function overloads in Python! (if you're feeling really nostalgic, have you seen [pointers.py](https://github.com/ZeroIntensity/pointers.py)?)

> From your perspective, is the first way a more natural way to call this?

Yeah, I think in a function like this where the first argument is the focus of the function and the rest of the args just modify the operation, my instinct is usually to make it positional.

>  What were your reservations about suggesting this initially?

My main concern is that the relevant args depend on the input type, which makes for a confusing docstring and signature. I think the only time I've seen single dispatch in use is in the Planetary Computer SDK where there's a [sign method](https://github.com/microsoft/planetary-computer-sdk-for-python/blob/30cc12bebe8a10bdd99bb0b32956cc2c8078a58d/planetary_computer/sas.py#L85) that dispatches for a bunch of different object types, but they all take the same args, so they avoid that complexity.

Reply:
> (if you're feeling really nostalgic, have you seen [pointers.py](https://github.com/ZeroIntensity/pointers.py)?)

OMG, that's hilarious (and frightening!)

> My main concern is that the relevant args depend on the input type, which makes for a confusing docstring and signature.

Definitely a warning to be heeded!  

As I noted above, I went ahead and implemented `@singledispatchmethod`s for `Raw.predict`, the overloaded `Transformed.predict`, and `Transformed.transform`.  I did take an initial stab at a docstring for both `Raw.predict` and `Transformed.transform`, which I leaned on ChatGPT to help create and get the expected formatting.  

I was hoping to reuse the docstring from `Raw.predict` for `Transformed.predict` with the use of `functools.wraps`, but I couldn't quite get it to work.  The dispatching doesn't seem to work nicely with `@wraps`.  For now, I don't have a docstring, but I'm guessing I should just copy it from `Raw.predict`?

Reply:
For now, I've copied the docstring directly from `Raw.predict` to `Transformed.predict`.